javascript/ql/src/experimental/Security/CWE-020/PostMessageNoOriginCheck.ql

```diff
@@ -0,0 +1,71 @@
+/**
+ * @name Missing `MessageEvent.origin` verification in `postMessage` handlers
+ * @description Missing the `MessageEvent.origin` verification in `postMessage` handlers, allows any windows to send arbitrary data to the `MessageEvent` listener.
+ * This could lead to unexpected behaviour, especially when `MessageEvent.data` is used in an unsafe way.
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @id js/missing-postmessageorigin-verification
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-20
+ */
+
+import javascript
+import semmle.javascript.security.dataflow.DOM
+
+/**
+ * A method call for the insecure functions used to verify the `MessageEvent.origin`.
+ */
+class InsufficientOriginChecks extends DataFlow::MethodCallNode {
+  InsufficientOriginChecks() {
+    exists(string name | name = getMethodName() |
+      name = "indexOf" or
+      name = "includes" or
+      name = "endsWith" or
+      name = "startsWith" or
+      name = "lastIndexOf"
+    )
+  }
+}
+
+/**
+ * A function handler for the `MessageEvent`.
+ */
+class PostMessageHandler extends DataFlow::FunctionNode {
+  PostMessageHandler() { exists(PostMessageEventHandler handler | this.getFunction() = handler) }
+}
+
+/**
+ * The `MessageEvent` parameter received by the handler
+ */
+class PostMessageEvent extends DataFlow::SourceNode {
+  PostMessageEvent() { exists(PostMessageHandler handler | this = handler.getParameter(0)) }
+
+  VarAccess event;
+  EqualityTest astNode;
+
+  /**
+   * Holds if an access on `MessageEvent.origin` is in an `EqualityTest` and there is no call of an insufficient verification method on `MessageEvent.origin`
+   */
+  predicate hasOriginChecked() {
+    exists(string prop | prop = "origin" or prop = "source" |
+      astNode.getAnOperand().(PropAccess).accesses(event, prop) and
+      event.mayReferToParameter*(this.asExpr()) and
+      not this.hasOriginInsufficientlyChecked()
+    )
+  }
+
+  /**
+   * Holds if there is an insufficient method call (i.e indexOf) used to verify `MessageEvent.origin`
+   */
+  predicate hasOriginInsufficientlyChecked() {
+    exists(InsufficientOriginChecks insufficientChecks |
+      this.getAPropertyRead("origin").getAMethodCall*() = insufficientChecks
+    )
```

Comment:
```suggestion
    this.getAPropertyRead("origin").getAMethodCall() instanceof InsufficientOriginChecks
```

Reply:
Thanks for the all the feedback. I really appreciated.

I have just a comment about this suggestion. Should it be 
```
this.getAPropertyRead("origin").getAMethodCall*() instanceof InsufficientOriginChecks
```
?

Without the `*` I don't catch for example the following case:
```
    let origin = event.origin.toLowerCase();
    let host = window.location.host;
    // BAD
    if (origin.startsWith(host) === -1)
        return;
```

This is because in the `if` statement the resolved predicate is `event.origin.toLowerCase().startsWith()` (please do not hesitate to correct me if I’m wrong).

Reply:
I see, that is a good point. Lets add the `*` then.

(If you are interested, then the way we would check for such extended flows in a non-experimental query, would be using "type tracking". Type tracking would also allow us to identify bad checks in other functions for instance. But that is an advanced feature that can be added later)

Reply:
Perfect, so I  leave the `*`.